python/perf-kernels/03-matrix-multiplication-stream-k.py

```diff
@@ -0,0 +1,407 @@
+            ACC_TYPE=ACC_TYPE,
+            GROUP_M=GROUP_M,
+            num_stages=num_stages,
+            num_warps=num_warps,
+        )
+        if matmul._debug:
+            print(f"{k1.n_regs} registers used, {k1.n_spills} spills")
+        k2 = full_tiles[(total_blocking_tiles,)](
+            a,
+            b,
+            c,
+            M,
+            N,
+            K,
+            a.stride(0),
+            a.stride(1),
+            b.stride(0),
+            b.stride(1),
+            c.stride(0),
+            c.stride(1),
+            total_tiles_streamk=total_tiles_streamk,
+            BLOCK_M=BLK_M,
+            BLOCK_N=BLK_N,
+            BLOCK_K=BLK_K,
+            ACC_TYPE=ACC_TYPE,
+            GROUP_M=GROUP_M,
+            num_stages=num_stages,
+            num_warps=num_warps,
+        )
+        if matmul._debug:
+            print(f"{k2.n_regs} registers used, {k2.n_spills} spills")
+        return c
+
+    @staticmethod
+    def forward(ctx, a: torch.Tensor, b: torch.Tensor, grid: int, BLK_M=128, BLK_N=128, BLK_K=32, two_tiles=True, num_stages=3, num_warps=4):
+        return matmul._call(a=a, b=b, total_programs_streamk=grid, BLK_M=BLK_M, BLK_N=BLK_N, BLK_K=BLK_K, two_tiles=two_tiles, num_warps=num_warps, num_stages=num_stages)
+
+
+# ---------------------------------------------------------------------------
+# Example and Benchmark
+# ---------------------------------------------------------------------------
+
+perf = lambda ms: 2 * m * n * k * 1e-12 / (ms * 1e-3)
+
+m, n, k = 8192, 8192, 8192  # some problem size to test
+A = torch.randn(m, k, device="cuda", dtype=torch.float16)
+B = torch.randn(k, n, device="cuda", dtype=torch.float16)
+
+matmul.set_debug(True)
+C = matmul.apply(A, B, total_sm, 128, 128, 32, 4, 4)
+matmul.set_debug(False)
+expected = A @ B
+
+assert torch.allclose(C, expected, atol=1), f"max: {(C - expected).abs().max().item()}\n{C}\n{expected}"
+
+# for debugging, uncomment the following line
+# exit(0)
+
+triton_ms = triton.testing.do_bench(lambda: torch.matmul(A, B))
+print(f"PyTorch: {triton_ms:.3f} ms  {perf(triton_ms):.3f} tflops")
+
+triton_ms = triton.testing.do_bench(lambda: matmul.apply(A, B, total_sm, 128, 128, 32, True, 4, 4))
+print(f"hybrid stream-k (grid={total_sm}): {triton_ms:.3f} ms  {perf(triton_ms):.3f} tflops")
+
+triton_ms = triton.testing.do_bench(lambda: matmul.apply(A, B, total_sm * 2, 128, 128, 32, True, 4, 4))
+print(f"hybrid stream-k (grid={total_sm * 2}): {triton_ms:.3f} ms  {perf(triton_ms):.3f} tflops")
+
+triton_ms = triton.testing.do_bench(lambda: matmul.apply(A, B, 0, 128, 128, 32, True, 4, 4))
+print(f"tile matmul (grid=0): {triton_ms:.3f} ms  {perf(triton_ms):.3f} tflops")
+
+
+
+exit(0)
+# ---------------------------------------------------------------------------
+# Log-sampled benchmark
+# ---------------------------------------------------------------------------
+
+# tried to reproduce the tests described in the paper
+num_samples = 1000  # 32768
```

Comment:
How would I be able to run such experiment on my end? I'd like to see if this large sampled benchmark performs better than the current GEMM kernels within Triton.